src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp

```diff
@@ -1187,6 +1187,16 @@ MergeTreeData::MutableDataPartPtr MergeTreeDataMergerMutator::mutatePartToTempor
     const ReservationPtr & space_reservation,
     TableLockHolder & holder)
 {
+    /// Mutator cannot be run in parallel with cleaner,
+    /// since cleaner may remove partial result of a merge,
+    /// and chances of this will be increase if:
+    /// - temporary_directories_lifetime is pretty low,
+    /// - someone adjust system-wide clock.
+    ///
+    /// And these will lead to missing columns in these part,
+    /// and the server will figure out that something is wrong only on DETACH/ATTACH.
+    std::unique_lock lock(data.clear_old_temporary_directories_mutex);
```

Comment:
But it may prevent cleanup thread to run for unbounded time. And also it will prevent concurrent mutations.

Reply:
Ok, done w/o lock (by introducing `MergeTreeDataMergerMutator::hasTemporaryPart()`)